pkg/payforput/payments.go

```diff
@@ -0,0 +1,211 @@
+package payforput
+
+import (
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/cashweb/keyserver/pkg/models"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/rs/zerolog/hlog"
+)
+
+// ValidatorFunc is the type of unction which can indicate if a request has
+// the appropriate headers to have been paid.
+type ValidatorFunc func(r *http.Request, secret string) bool
+
+// PaymentEnforcer ensures that the request has been paid for properly by
+// checking for a payment authorization. If there is no payment authorization
+// header, then we send a BIP70 payment request
+type PaymentEnforcer struct {
+	// Location to where we should redirect payments, and has the PaymentHandler
+	// installed
+	PaymentURL string
+	// Validator is a func that validates that a request has a valid proof of
+	// payment
+	Validator ValidatorFunc
+	// Secret is the HMAC secret used for generating and validating tokens
+	// NOTE: This may be swapped out at a later date.
+	Secret string
+}
+
+// New returns a new payment enforcer that can be used for easy BIP70 integration
+// for the keyserver.
+func New(PaymentURL string, Validator ValidatorFunc) *PaymentEnforcer {
+	pe := &PaymentEnforcer{
+		PaymentURL: PaymentURL,
+		Validator:  Validator,
+		Secret:     "",
+	}
+	if pe.Validator == nil {
+		pe.Validator = DefaultValidator
+	}
+	// Generate an ephemeral secret and hold on to it
+	if pe.Secret == "" {
+		pe.Secret = RandString(64)
+	}
+
+	return pe
+}
+
+// DefaultValidator is the default request payment validator
+func DefaultValidator(r *http.Request, secret string) bool {
+	// First attempt to get the code from the querystring
+	token := r.URL.Query().Get("code")
+
+	headerToken := r.Header.Get("Authorization")
+	// If we had an Authorization header, use that instead.
+	if headerToken != "" && len(headerToken) >= 4 && headerToken[0:4] == "POP " {
+		token = headerToken[4:]
+	}
+
+	// Remove the code query param, as it wasn't part of the HMAC hash
+	url := *r.URL
+	url.RawQuery = ""
+	// Validate that the HMAC is valid for this URL.
+	return ValidateHMACToken(url.String(), token, secret)
+}
+
+// PaymentHandler is an http handler that implements a check for payment,
+// along with a redirect to the original location with the key
+func (e *PaymentEnforcer) PaymentHandler(w http.ResponseWriter,
+	r *http.Request) {
+	log := hlog.FromRequest(r)
+	defer r.Body.Close()
+
+	// Check BIP70 headers
+	contentType := r.Header.Get("Content-Type")
+	if contentType != "application/bitcoin-payment" {
+		http.Error(w, "invalid content type", http.StatusUnsupportedMediaType)
+		return
+	}
+	acceptHeader := r.Header.Get("Accept")
+	if acceptHeader != "application/bitcoin-paymentack" {
+		http.Error(w, "invalid content type", http.StatusUnsupportedMediaType)
```

Comment:
I think this guy is supposed to be 406 NotAcceptable?

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/406